### DISCUSSION TITLE
Nullable Sorting always at the end?

### DISCUSSION THREAD

**Discussion by SbiCA** [ravendb/ravendb#16703] [Q&A] [answered]
Hi all,

I was wondering if there is a possibility to define where `null` should be in terms of ordering by a field. In our case we have a `TimeTaken` that is not always applicable but we'd like the post without it to be at the end no matter if the sorting is `asc` or `desc`. Is anyone aware of something like that?

```c# 

public class Post {
   string Id {get;set;}
   // it's not always present
   DateTime? TimeTaken  {get;set;}
}

```

Current behaviour is (using an index of course)

``` rql
from index 'AllPostsBy/TimeTaken'
order by TimeTaken desc
```
descending, the null values are at the end.
ascending, the null values are at the beginning.

**Comment by ayende** [comment]
Easiest way to handle that, in the index, use:

`TimeTaken = p.TimeTaken ?? TimeSpan.MaxValue`

**Comment by SbiCA** [comment]
@ayende  thanks for the quick response I'm afraid I doesn't solve my problem (Probably I was too vague in my description).

I added a small demo project, the asc tests succeeds but the desc one fails. 

https://github.com/SbiCA/RavenSorting/blob/ce1b41663f1dd530e515a0babe25461572655778/SortingWithNullTests.cs#L39 

I'm aware that it's tricky because with switching the order you want to have null in different places

**Reply by SbiCA** [reply]
I seems that it would require a custom sorter since from what I can see in the the example here https://ravendb.net/articles/ravendb-custom-sorting-and-when-not-to-use-it 

``` c#
from candidate in docs.Candidates
select new 
{
  candidate.Name,
  candidate.Education,
  EducationForSorting = candidate.Education.ToString() switch
    {
        "HighSchool" => 12,
        "Associate"  => 14,
        "Bachelor"   => 16,
        "Master"     => 18,
        "Doctor"     => 22,
        _            =>  Int.MinValue / Int.MaxValue // this would need to switch based on asc or desc
    }
}

```

**Reply by SbiCA** [reply]
maybe to give the requirement perspective, it's to show posts without `TimeTaken` always at the end since you don't have the data

**Comment by ayende** [comment, marked as answer]
In your index, use:

```
HasTimeTaken = p.TimeTaken != null ?  0 : 1
TimeTaken = p.TimeTaken
```

And in your query:

`order by HasTimeTaken asc, TimeTaken desc`

In this way, the one without `HasTimeTaken` will be at the end.

**Reply by SbiCA** [reply]
@ayende  great that works 🎉  small correction 

`order by HasTimeTaken desc, TimeTaken desc`

I was wondering about the additional index space it would add, since this would be a field more on the index (the collection has a few million docs)🤔  I guess the only way to avoid that would be to use the custom sorter right?

**Reply by ayende** [reply]
I would be very surprised if this is very costly in terms of disk usage.
And custom sorter is about 10x level more complex to build.

**Reply by SbiCA** [reply]
Agreed, that's actually an elegant way to solve it. 

Btw. you were right there isn't a noticeable difference in terms of disk usage, thx again!
